scapy/contrib/automotive/xcp/xcp.py

```diff
@@ -0,0 +1,463 @@
+    DAQEventInfoPositiveResponse, ProgramStartPositiveResponse, \
+    PgmProcessorPositiveResponse, SectorInfoPositiveResponse
+from scapy.contrib.automotive.xcp.utils import get_timestamp_length, \
+    identification_field_needs_alignment, get_daq_length, \
+    get_daq_data_field_length
+from scapy.fields import ByteEnumField, ShortField, XBitField, \
+    FlagsField, ByteField, ThreeBytesField, StrField, ConditionalField, \
+    XByteField, StrLenField
+from scapy.layers.can import CAN
+from scapy.layers.inet import UDP, TCP
+from scapy.packet import Packet, bind_layers, bind_bottom_up, bind_top_down
+
+if "XCP" not in conf.contribs:
+    conf.contribs["XCP"] = {}
+
+# 0 stands for Intel/little-endian format, 1 for Motorola/big-endian format
+conf.contribs["XCP"]["byte_order"] = 1
+conf.contribs["XCP"]["allow_byte_order_change"] = True
+conf.contribs["XCP"]["Address_Granularity_Byte"] = None  # Can be 1, 2 or 4
+conf.contribs["XCP"]["allow_ag_change"] = True
+
+conf.contribs["XCP"]["MAX_CTO"] = None
+conf.contribs["XCP"]["MAX_DTO"] = None
+conf.contribs["XCP"]["allow_cto_and_dto_change"] = True
+
+conf.contribs['XCP']['timestamp_size'] = 0
+
+
+# Specifications from:
+# http://read.pudn.com/downloads293/doc/comm/1316424/ASAM_XCP_Part1-Overview_V1.0.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%202-%20Protocol%20Layer%20Specification%20-1.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%203-%20Transport_layer_specification_xcp_on_can_1-0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%204-%20Interface%20Specification%20-1.0.pdf # noqa: E501
+# http://read.pudn.com/downloads192/doc/comm/903802/XCP%20-Part%205-%20Example%20Communication%20Sequences%20-1.0.pdf # noqa: E501
+
+# XCP on USB is left out because it has "no practical meaning"
+# XCP on Lin is left out because it has no official specification
+class XCPOnCAN(CAN):
+    name = "Universal calibration and measurement protocol on CAN"
+    fields_desc = [
+        FlagsField("flags", 0, 3, ["error",
+                                   "remote_transmission_request",
+                                   "extended"]),
+        XBitField("identifier", 0, 29),
+        ByteField("length", None),
+        ThreeBytesField("reserved", 0),
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.length is None:
+            tmp_len = len(pay)
+            pkt = pkt[:4] + struct.pack("B", tmp_len) + pkt[5:]
+
+        return super(XCPOnCAN, self).post_build(pkt, pay)
+
+    def extract_padding(self, p):
+        return p, None
+
+
+class XCPOnUDP(UDP):
+    name = "Universal calibration and measurement protocol on Ethernet"
+    fields_desc = UDP.fields_desc + [
+        ShortField("length", None),
+        ShortField("ctr", 0),  # counter
+    ]
+
+    def post_build(self, pkt, pay):
+        if self.length is None:
+            tmp_len = len(pay)
```

Comment:
Could you please add an unit test which checks this branch.

Reply:
@akorb I wrote a unittest, however this does only change the bytestring and not the length field.
If the length field should be changed too, we should add it here. (Same for XCPOnTCP)